### PR TITLE
retain_graph=True (optional param in backpack)

### LINKED ISSUE #213: Second order gradient for Batch-wise HVP
Hi,

Thanks for this great library. I am wondering how can we compute the gradient of sum of gradient?

I am trying to implement the Hessian-Vector-Product (HVP) with the following code:

```
def batch_hvp(self, model, loss, params_list, batch_grad_list):
    if len(params_list) != len(batch_grad_list):
        raise (ValueError("w and v must have the same length."))

    one_sample_grad_list = grad(loss, params_list, retain_graph=True, create_graph=True)

    elemwise_products = 0
    for grad_elem, v_elem in zip(one_sample_grad_list, batch_grad_list):
        sum_over_dims = []
        for i in range(len(v_elem.shape)):
            sum_over_dims.append(i)
        sum_over_dims = tuple(sum_over_dims[1:])
        elemwise_products += torch.sum(grad_elem.unsqueeze(0) * v_elem.detach(), sum_over_dims)

    with backpack(BatchGrad()):
        elemwise_products.backward()   # problem: has no attribute 'input0'
        return_grads = [p.grad_batch for p in model.parameters() if p.requires_grad]

    return return_grads
```
I encounter the  " has no attribute 'input0'  " problem when I call the backward(), is it possible to get batch gradients for return_grads?

For now, I am only using the for loop to compute the gradient.

```
def batch_hvp(self, model, loss, params_list, batch_grad_list):
    if len(params_list) != len(batch_grad_list):
        raise (ValueError("w and v must have the same length."))

    one_sample_grad_list = grad(loss, params_list, retain_graph=True, create_graph=True)

    elemwise_products = 0
    for grad_elem, v_elem in zip(one_sample_grad_list, batch_grad_list):
        sum_over_dims = []
        for i in range(len(v_elem.shape)):
            sum_over_dims.append(i)
        sum_over_dims = tuple(sum_over_dims[1:])
        elemwise_products += torch.sum(grad_elem.unsqueeze(0) * v_elem.detach(), sum_over_dims)
    
    # The for-loop version 
    grad_cache = []
    for i in range(elemwise_products.shape[0]):
        elemwise_products[i].backward(retain_graph=True)
        grad_cache.append([p.grad.clone() for p in model.parameters() if p.requires_grad])
    grad_cache = list(zip(*grad_cache))
    return_grads = []
    for l_id in range(len(grad_cache)):
        return_grads.append(torch.cat([g.unsqueeze(0) for g in grad_cache[l_id]], dim=0))

    return return_grads
```

Thanks in advance!

### PR DESCRIPTION
**Pull request by schaefertim**
Handle the case where retain_graph=True.
This fixes #213
and fixes first half of fKunstner/backpack-discuss#115

This is the alternative to #214.

Here, an additional parameter is added to `backpack()`. Retain graph can be used with
```Python
with backpack(retain_graph=True):
  loss.backward(retain_graph=True)
```

### CODE REVIEW

File: backpack/__init__.py [hunk 1]
```diff
@@ -36,6 +37,9 @@ def __init__(
                 all BackPACK extensions have run. Takes a ``torch.nn.Module`` and returns
                 ``None``. Default: ``None`` (no operation will be performed).
             debug: Print debug messages during the backward pass. Default: ``False``.
+            retain_graph: Determines whether BackPack quantities should be kept for an
+                additional backward pass. Usually, must have same value
+                as the argument ``retain_graph`` in ``backward()``.
```
Comment (f-dangel) [inline backpack/__init__.py:42]: Missing default value

File: backpack/__init__.py [hunk 2]
```diff
@@ -205,7 +213,7 @@ def hook_run_extensions(
         print("[DEBUG] Running extension hook on", module)
     CTX.get_extension_hook()(module)
 
-    if not (
+    if not CTX.get_retain_graph() and not (
```
Comment (f-dangel) [inline backpack/__init__.py:216]: write `not a and not b` as `not (a or b)`

File: test/test_retain_graph.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    with raises(AssertionError):
+        _clear_input_output(model)
+
+    # finally doing a normal backward pass that verifies graph is clear again
+    with backpack(BatchGrad()):
+        loss.backward()
+    _clear_input_output(model)
+
+
+def _clear_input_output(parent_module: Module) -> bool:
```
Comment (f-dangel) [inline test/test_retain_graph.py:43]: Docstring

File: test/test_retain_graph.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    if list(parent_module.children()):
+        return all([_clear_input_output(module) for module in parent_module.children()])
+    elif hasattr(parent_module, "input0") or hasattr(parent_module, "output"):
+        raise AssertionError(
+            f"graph should be clear, but {parent_module} has input0 or output."
+        )
+    else:
+        return True
```
Comment (f-dangel) [inline test/test_retain_graph.py:51]: Does this function have to return anything? It seems to me that it performs a check and raises an exception if the check fails.

File: test/test_retain_graph.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    # finally doing a normal backward pass that verifies graph is clear again
+    with backpack(BatchGrad()):
+        loss.backward()
+    _clear_input_output(model)
+
+
+def _clear_input_output(parent_module: Module) -> bool:
+    if list(parent_module.children()):
+        return all([_clear_input_output(module) for module in parent_module.children()])
```
Comment (f-dangel) [inline test/test_retain_graph.py:45]: extract `parent_module.children()`

File: test/test_retain_graph.py (hunk 3 above)
Comment (f-dangel) [inline test/test_retain_graph.py:43]: Rename into `_check_clear_io`

File: test/test_retain_graph.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    elif hasattr(parent_module, "input0") or hasattr(parent_module, "output"):
```
Comment (f-dangel) [inline test/test_retain_graph.py:46]: `any(hasattr(parent_module, a) for a in ["input0", "output"])`

Maybe extract the list and re-use in exception message.

File: test/test_retain_graph.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+from backpack import backpack, extend
+from backpack.extensions import BatchGrad
+
+
+def test_retain_graph():
+    model = extend(Sequential(Linear(4, 6), Linear(6, 5)))
+    loss_fn = extend(CrossEntropyLoss())
+
+    # after a forward pass graph is not clear
+    loss = loss_fn(model(rand(8, 4)), randint(5, (8,)))
```
Comment (f-dangel) [inline test/test_retain_graph.py:15]: extract input and labels, fix random seed.

File: backpack/__init__.py (hunk 1 above)
Comment (schaefertim) [reply backpack/__init__.py:42]: Added.

File: backpack/__init__.py (hunk 2 above)
Comment (schaefertim) [reply backpack/__init__.py:216]: Done.

File: test/test_retain_graph.py (hunk 3 above)
Comment (schaefertim) [reply test/test_retain_graph.py:43]: added

File: test/test_retain_graph.py (hunk 4 above)
Comment (schaefertim) [reply test/test_retain_graph.py:51]: Changed it to return None

File: test/test_retain_graph.py (hunk 5 above)
Comment (schaefertim) [reply test/test_retain_graph.py:45]: outdated comment, since code changed significantly

File: test/test_retain_graph.py (hunk 3 above)
Comment (schaefertim) [reply test/test_retain_graph.py:43]: Yes. Better name.

File: test/test_retain_graph.py (hunk 6 above)
Comment (schaefertim) [reply test/test_retain_graph.py:46]: outdated. code changed significantly

File: test/test_retain_graph.py (hunk 7 above)
Comment (schaefertim) [reply test/test_retain_graph.py:15]: Fixed the see.
What do you mean by "extract input and labels"?

Comment (f-dangel) [reply test/test_retain_graph.py:21]:
```
inputs = rand(8, 4)
labels = randint(5, (8,))
```
Then you can re-use them further below

Comment (schaefertim) [reply test/test_retain_graph.py]: Ok, implemented it.

File: docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+BATCH_SIZE = 32
+DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+torch.manual_seed(0)
+
+
+# %%
+# Let's define a function that calculates individual hessian_vector products.
+# This function was provided by Haonan Wang (github: haonan3) in #213.
+def batch_hvp(model, loss, params_list, batch_grad_list):
```
Comment (f-dangel) [inline docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py:22]: Docstring, with shape conventions.

File: docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+batch_hvp = batch_hvp(model, loss, params_list, batch_grad_list)
+
+print("\nresult")
+for result in batch_hvp:
+    print(result.shape)
+
+
+# %%
+# Compare the result with the for-loop version
+def batch_hvp_for_loop(model, loss, params_list, batch_grad_list):
```
Comment (f-dangel) [inline docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py:88]: Docstring, with shapes.

File: docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+# Let's define a function that calculates individual hessian_vector products.
+# This function was provided by Haonan Wang (github: haonan3) in #213.
+def batch_hvp(model, loss, params_list, batch_grad_list):
+    if len(params_list) != len(batch_grad_list):
+        raise (ValueError("w and v must have the same length."))
+
+    with backpack(retain_graph=True):
+        one_sample_grad_list = grad(
+            loss, params_list, retain_graph=True, create_graph=True
+        )
```
Comment (f-dangel) [inline docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py:29]: I think it would be better to replace this with a `loss = forward(); loss.backward()` which is closer to how BackPACK is supposed to be used.

File: docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        sum_over_dims = []
+        for i in range(len(v_elem.shape)):
+            sum_over_dims.append(i)
+        sum_over_dims = tuple(sum_over_dims[1:])
+        elemwise_products += torch.sum(
+            grad_elem.unsqueeze(0) * v_elem.detach(), sum_over_dims
+        )
+
+    with backpack(BatchGrad()):
+        elemwise_products.backward()  # problem: has no attribute 'input0'
```
Comment (f-dangel) [inline docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py:42]: is the comment still relevant?

File: docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py (hunk 9 above)
Comment (f-dangel) [reply docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py:88]: I suspect you need access to the `input` and `labels` to implement per-sample Hessian-vector products.

File: docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+    if fail:
+        raise Exception("At least one IHVP does not match.")
+
+
+if __name__ == "__main__":
+    print("\nWorking model (parameters appear once in graph)")
+    compare_batch_hvp(make_working_model)
+    print("\nFailing model (weight of 2nd layer appears multiple times in graph)")
+    compare_batch_hvp(make_failing_model)
```
Comment (f-dangel) [inline docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py:106]: Remove

File: test/test_retain_graph.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def test_for_loop_replace() -> None:
+    """Application of retain_graph: replace an outer for-loop.
+
+    This test is based on issue #220 opened by Romain3Ch216.
+    """
+    B = 20
+    M = 10
+    h = 2
```
Comment (f-dangel) [inline test/test_retain_graph.py:77]: Reduce sizes for faster execution. Set random seed.

File: test/test_retain_graph.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+    x = randn(B, h)
+    fc = Linear(h, M)
+    fc = extend(fc)
```
Comment (f-dangel) [inline test/test_retain_graph.py:81]: merge

File: backpack/__init__.py [hunk 15]
```diff
@@ -39,6 +40,9 @@ def __init__(
                 all BackPACK extensions have run. Takes a ``torch.nn.Module`` and returns
                 ``None``. Default: ``None`` (no operation will be performed).
             debug: Print debug messages during the backward pass. Default: ``False``.
+            retain_graph: Determines whether BackPack quantities should be kept for an
```
Comment (f-dangel) [inline backpack/__init__.py:43]: replace 'BackPack quantities' with `BackPACK IO`?

File: test/test_retain_graph.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    for child_module in module.children():
+        _check_no_io(child_module)
+
+    io_strs = ["input0", "output"]
+    if any(hasattr(module, io) for io in io_strs):
+        raise AssertionError(f"IO should be clear, but {module} has one of {io_strs}.")
+
+
+def test_for_loop_replace() -> None:
```
Comment (f-dangel) [inline test/test_retain_graph.py:70]: Improve name

File: test/test_retain_graph.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    """Application of retain_graph: replace an outer for-loop.
+
+    This test is based on issue #220 opened by Romain3Ch216.
```
Comment (f-dangel) [inline test/test_retain_graph.py:73]: Add one more sentence what the issue wants to acomplish

Comment (schaefertim) [reply test/test_retain_graph.py:73]: I thought the whole point of the issue is to replace one for-loop with BatchGrad.
Therefore I chose this name and description.
What do I miss?

Comment (f-dangel) [reply test/test_retain_graph.py:73]: Added a description what the issue aims to do.
